JIT: fix patchpoint offset encoding

AndyAyersMS opened the pull request:
We can now have Tier0 locals at byte offsets, so rework how the offset
information is incoded in patchpoints to make this possible.

Closes #68194.

AndyAyersMS commented:
@jakobbotsch PTAL
cc @dotnet/jit-contrib 

Current SPMI data (if there is any) will still "work". Will kick off a new ASP.NET collection once this lands.

jakobbotsch commented on src/coreclr/inc/patchpointinfo.h:
```diff
@@ -157,12 +157,12 @@ struct PatchpointInfo
     void SetOffset(unsigned localNum, int offset)
     {
-        m_offsetAndExposureData[localNum] = offset;
+        m_offsetAndExposureData[localNum] = (offset << OFFSET_SHIFT);
```
Seems like this will clear out the exposure bit.

AndyAyersMS replied:
It's always been that way and calling `SetIsExposed` was optional. I'll rework this so they get set together.

AndyAyersMS committed:
JIT: fix patchpoint offset encoding

We can now have Tier0 locals at byte offsets, so rework how the offset
information is incoded in patchpoints to make this possible.

Closes #68194.

AndyAyersMS commented:
Revised (and merged/force-pushed to fix up the committer email).
